python/plugins/processing/algs/qgis/SplitWithLines.py

```diff
@@ -81,107 +81,111 @@ def processAlgorithm(self, progress):
-
-                                            for aNewGeom in newGeometries:
-                                                inGeoms.append(aNewGeom)
-                                    else:
-                                        QgsMessageLog.logMessage("appending else")
+                            if split_geom_engine.intersects(inGeom.geometry()):
+                                inPoints = vector.extractPoints(inGeom)
+                                if splitterPList == None:
+                                    splitterPList = vector.extractPoints(splitGeom)
+
+                                try:
+                                    result, newGeometries, topoTestPoints = inGeom.splitGeometry(splitterPList, False)
+                                except:
+                                    ProcessingLog.addToLog(ProcessingLog.LOG_WARNING,
+                                                           self.tr('Geometry exception while splitting'))
+                                    result = 1
+
+                                # splitGeometry: If there are several intersections
+                                # between geometry and splitLine, only the first one is considered.
+                                if result == 0:  # split occurred
+                                    if inPoints == vector.extractPoints(inGeom):
+                                        # bug in splitGeometry: sometimes it returns 0 but
+                                        # the geometry is unchanged
                                         outGeoms.append(inGeom)
+                                    else:
+                                        inGeoms.append(inGeom)
+
+                                        for aNewGeom in newGeometries:
+                                            inGeoms.append(aNewGeom)
                                 else:
                                     outGeoms.append(inGeom)
+                            else:
+                                outGeoms.append(inGeom)
 
-                            inGeoms = outGeoms
+                        inGeoms = outGeoms
 
-                for aGeom in inGeoms:
-                    passed = True
+            parts = []
 
-                    if QgsWkbTypes.geometryType( aGeom.wkbType() )  == QgsWkbTypes.LineGeometry \
-                            and not QgsWkbTypes.isMultiType(aGeom.wkbType()):
-                        passed = len(aGeom.asPolyline()) > 2
+            for aGeom in inGeoms:
+                passed = True
 
-                        if not passed:
-                            passed = (len(aGeom.asPolyline()) == 2 and
-                                      aGeom.asPolyline()[0] != aGeom.asPolyline()[1])
-                            # sometimes splitting results in lines of zero length
+                if QgsWkbTypes.geometryType( aGeom.wkbType() )  == QgsWkbTypes.LineGeometry:
+                    firstVertex = aGeom.vertexAt(0)
```

Comment:
This could be simplified to `passed = aGeom.geometry().numPoints() > 2`. It also avoids the possibility that the third vertex is validly located at 0,0.